### DISCUSSION TITLE
Fallback for Option?

### DISCUSSION THREAD

**Discussion by failable** [pacak/bpaf#310] [Q&A] [answered]
Why can't I set a fallback value for `Option<String>` type in this case?

```rs
use bpaf::Bpaf;

#[allow(dead_code)]
#[derive(Debug, Clone, Bpaf)]
#[bpaf(options)]
pub struct Options {
    #[bpaf(long, fallback(Some(String::from("main"))))]
    revision: Option<String>,

    #[bpaf(long, fallback(String::from("main")))]
    revision2: String,
}

fn main() {
    println!("{:?}", options().run())
}
```

**Comment by pacak** [comment, marked as answer]
First let's take a look at the code that gets generated in both cases:

```rust
let revision = long("revision").argument::<String>("ARG").optional().fallback(Some(String::from("main")));
let revision2 = long("revision2").argument::<String>("ARG").fallback(String::from("main"));
```

Both `fallback` and `optional` work by catching "item not found" error and converting it into a value - specified value in case of `fallback` and `None` in case of `optional`. Later additionally wraps present value with `Some`.

- https://docs.rs/bpaf/0.9.5/bpaf/trait.Parser.html#method.optional
- https://docs.rs/bpaf/0.9.5/bpaf/trait.Parser.html#method.fallback

If `revision` is always will be present either from user input or from the fallback - does it make sense to have it `Option`? If you are planning to use this datatype somewhere else and it needs to be `Option` because of that - you can do something like this:

```rust
use bpaf::Bpaf;

#[allow(dead_code)]
#[derive(Debug, Clone, Bpaf)]
#[bpaf(options)]
pub struct Options {
    #[bpaf(long, argument, fallback(String::from("main")), map(Some))]
    revision: String,
}

fn main() {
    println!("{:?}", options().run())
}
```
